posix/fdtable.c

```diff
@@ -0,0 +1,245 @@
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
+
+    if (new_size > _table_size)
+    {
+        entry_t* p;
+        size_t n = new_size;
+
+        /* Reallocate the table. */
+        if (!(p = oe_realloc(_table, n * sizeof(entry_t))))
+            goto done;
+
+        /* Zero-fill the unused portion. */
+        {
+            const size_t num_bytes = (n - _table_size) * sizeof(entry_t);
+
+            if (oe_memset_s(p + _table_size, num_bytes, 0, num_bytes) != OE_OK)
+                goto done;
+        }
+
+        _table = p;
+        _table_size = new_size;
+    }
+
+    ret = 0;
+
+done:
+    return ret;
+}
+
+/*
+**==============================================================================
+**
+** Public interface:
+**
+**==============================================================================
+*/
+
+int oe_fdtable_assign(oe_device_t* device)
+{
+    int ret = -1;
+    size_t index;
+
+    oe_spin_lock(&_lock);
+
+    if (!device)
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    /* The file-descriptor table must be big enough for standard devices. */
+    if (_resize_table(TABLE_CHUNK_SIZE) != 0)
+        OE_RAISE_ERRNO(OE_ENOMEM);
+
+    /* Search for a free slot in the file descriptor table. */
+    for (index = OE_STDERR_FILENO + 1; index < _table_size; index++)
+    {
+        if (!_table[index])
+            break;
+    }
+
+    /* If no free slot found, expand size of the file descriptor table. */
+    if (index == _table_size)
+    {
+        if (_resize_table(_table_size + 1) != 0)
+            OE_RAISE_ERRNO(OE_ENOMEM);
+    }
+
+    _table[index] = device;
+    ret = (int)index;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_fdtable_release(int fd)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    /* Fail if fd is out of range. */
+    if (!(fd >= 0 && (size_t)fd < _table_size))
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    /* Fail if entry was never assigned. */
+    if (!_table[fd])
+        OE_RAISE_ERRNO(OE_EINVAL);
+
+    _table[fd] = NULL;
+
+    ret = 0;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+int oe_fdtable_reassign(int fd, oe_device_t* device)
+{
+    int ret = -1;
+
+    oe_spin_lock(&_lock);
+
+    /* Make table big enough to contain this file-descriptor. */
+    if (fd >= 0)
+        _resize_table((size_t)fd + 1);
+
+    if (fd < 0 || (size_t)fd >= _table_size)
+        OE_RAISE_ERRNO(OE_EBADF);
+
+    if (_table[fd])
+    {
+        if (OE_CALL_BASE(close, _table[fd]) != 0)
+            OE_RAISE_ERRNO(oe_errno);
+    }
+
+    /* Set the device. */
+    _table[fd] = device;
+
+    ret = 0;
+
+done:
+
+    oe_spin_unlock(&_lock);
+
+    return ret;
+}
+
+static oe_device_t* _get_fd_device(int fd)
+{
+    oe_device_t* ret = NULL;
+
+    /* Auto-load the console file system. */
+    if (oe_load_module_console_file_system() != OE_OK)
+        OE_RAISE_ERRNO(oe_errno);
```

Comment:
This works but there's an ordering dependency that probably should be called out or preferably avoided: you could call reassign to one of the standard interfaces before the first get call, which would overwrite the assignment.

This is not a possible as far as I can tell in the current changeset since `oe_dup`, `oe_dup2`, `oe_close` all hit get before an assignment to standard interfaces can occur, but  folks need to know to maintain that.

Reply:
Thanks Simon. I added the **_initialize()** function which sets up stdin, stdout, and stderr to all public functions in fdtable.c.